File: src/Extensibility/SharpDetect.Plugins/Deadlock/Descriptors/ArgumentTypeDescriptor.cs
// Copyright 2025 Andrej Čižmárik and Contributors
// SPDX-License-Identifier: Apache-2.0

using SharpDetect.Core.Events.Profiler;

namespace SharpDetect.Plugins.Deadlock.Descriptors;

public record ArgumentTypeDescriptor
{
    public CorElementType[] ElementTypes { get; init; }
    public string? TypeName { get; init; }

    private ArgumentTypeDescriptor(
        CorElementType[] elementTypes,
        string? typeName = null)
    {
        ElementTypes = elementTypes;
        TypeName = typeName;
    }
    
    public static ArgumentTypeDescriptor CreateSimple(CorElementType elementType) 
        => new([elementType]);
    
    public static ArgumentTypeDescriptor CreateClass(string typeName) 
        => new([CorElementType.ELEMENT_TYPE_CLASS], typeName);
    
    public static ArgumentTypeDescriptor CreateValueType(string typeName) 
        => new([CorElementType.ELEMENT_TYPE_VALUETYPE], typeName);
    
    public static ArgumentTypeDescriptor CreateByRef(ArgumentTypeDescriptor innerType) 
        => new([CorElementType.ELEMENT_TYPE_BYREF, .. innerType.ElementTypes], innerType.TypeName);
}

File: src/SharpDetect.Profiler/SharpDetect.Concurrency.Profiler/ArgumentTypeDescriptor.cpp
// Copyright 2026 Andrej Čižmárik and Contributors
// SPDX-License-Identifier: Apache-2.0

#include "ArgumentTypeDescriptor.h"

void Profiler::to_json(nlohmann::json& json, const ArgumentTypeDescriptor& descriptor)
{
    json["elementTypes"] = descriptor.elementTypes;
    
    if (!descriptor.typeName.has_value())
        json["typeName"] = descriptor.typeName.value();
}

void Profiler::from_json(const nlohmann::json& json, ArgumentTypeDescriptor& descriptor)
{
    descriptor.elementTypes = json.at("elementTypes").get<std::vector<CorElementType>>();

	auto const typeNameIt = json.find("typeName");

	if (typeNameIt != json.cend() && !(*typeNameIt).is_null())
		descriptor.typeName = json.at("typeName");
}